whylog/config/rule.py

```diff
@@ -57,6 +64,140 @@ def get_causes_parsers(self):
+
+    def _is_primary_key_constraint(self, clues_groups):
+        base_parser_number = clues_groups[1][0]
+        depended_parser_number = clues_groups[0][0]
+        base_parser_group_number = clues_groups[1][1]
+        depended_group_number = clues_groups[0][1]
+        return self._is_primary_key_group(base_parser_group_number, base_parser_number) and \
+               self._is_primary_key_group(depended_group_number, depended_parser_number)
+
+    def _calculate_parser_bounds(self, base_parser_number, params, group_type, parser_ranges):
+        max_delta = params.get('max_delta')
+        min_delta = params.get('min_delta')
+        left_bound, right_bound = self._get_base_bounds(
+            base_parser_number, group_type, parser_ranges
+        )
+        converter = CONVERTION_MAPPING[group_type]
+        new_left_bound = converter.switch_by_delta(left_bound, max_delta, "max")
+        new_right_bound = converter.switch_by_delta(right_bound, min_delta, "min")
+        return {group_type: {"left_bound": new_left_bound, "right_bound": new_right_bound}}
+
+    @classmethod
+    def _get_base_bounds(cls, base_parser_number, group_type, parser_ranges):
+        left_bound = parser_ranges[base_parser_number][group_type]["left_bound"]
+        right_bound = parser_ranges[base_parser_number][group_type]["right_bound"]
+        return left_bound, right_bound
+
+    def _aggregate_by_log_type(self, parser_ranges):
+        search_ranges = {}
+        for parser_number, ranges in six.iteritems(parser_ranges):
+            parser_log_type = self._causes[parser_number - 1].log_type
+            if parser_log_type not in search_ranges:
+                search_ranges[parser_log_type] = ranges
+                continue
+            self._update_log_type_ranges(parser_log_type, ranges, search_ranges)
+        return search_ranges
+
+    def _update_log_type_ranges(self, parser_log_type, ranges, search_ranges):
+        for group_type in ranges:
+            if group_type not in search_ranges[parser_log_type]:
+                search_ranges[parser_log_type][group_type] = ranges[group_type]
+                continue
+            self._update_bounds(search_ranges[parser_log_type][group_type], ranges[group_type])
+
+    @classmethod
+    def _update_bounds(cls, old_bounds_dict, parser_bound_dict):
+        old_bounds_dict["left_bound"] = min(
+            old_bounds_dict["left_bound"], parser_bound_dict["left_bound"]
+        )
+        old_bounds_dict["right_bound"] = max(
+            old_bounds_dict["right_bound"], parser_bound_dict["right_bound"]
+        )
+
+    def _is_primary_key_group(self, parser_group_number, parser_number):
+        if parser_number == self.EFFECT_NUMBER:
+            return self._effect.is_primary_key(parser_group_number)
+        return self._causes[parser_number - 1].is_primary_key(parser_group_number)
+
+    def _aggregate_constraints(self):
+        # Aggregate constraints have to have min_delta and max_delta params
```

Comment:
Could you please also describe what this function really do instead of describing only its assumptions?